Divide portage_utils.open() into two distinct implementations for Python 2.7 and 3

joanise opened the pull request:
The Python 2 version is like open in Python 2.7, with no encoding or newline option. The caller has to deal with that themselves if they want to. open() only takes care of the magic of files/compressed files/pipes/stdin/stdout.

The Python 3 version is enriched with features that Python 3's open has: an encoding parameters, and a newline parameter. Those now have consistent behaviour, no matter how the file is opened.

In both cases, we strive to have an intuitive implementation, staying close to what the given version of Python would do.

Exception: in Python 3, when encoding= is not specified, we default to "utf8", unlike builtins.open(). While less python3-ic because good Python 3 programming requires providing the encoding on every darn open() call you make, it makes code more portable, because on Linux "utf8" ends up being the default most of the time, so testing rarely flags the problem, while on Windows "utf8" is not typically the default so we end up with code that's broken on Windows.

joanise committed:
patches for proper utf-8 handling with Python 3.6

joanise committed:
clean_utf8 test: make sure the output is the same with stdin or a file

joanise committed:
better utf8 testing

joanise committed:
portage_utils now divides open() between python2 and python3

The Python 2 version is like open in Python 2.7, with no encoding or
newline option. The caller has to deal with that themselves if they want
to. open() only takes care of the magic of files/compressed
files/pipes/stdin/stdout.

The Python 3 version is enriched with features that Python 3's open has:
an encoding parameters, and a newline parameter. Those now have
consistent behaviour, no matter how the file is opened.

In both cases, we strive to have an intuitive implementation, staying
close to what the given version of Python would do.

Exception: in Python 3, when encoding= is not specified, we default to
"utf8", unlike builtins.open(). While less python3-ic because good
Python 3 programming requires providing the encoding on every darn
open() call you make, it makes code more portable, because on Linux
"utf8" ends up being the default most of the time, so testing rarely
flags the problem, while on Windows "utf8" is not typically the default
so we end up with code that's broken on Windows.

joanise committed:
Added new temporary files to TEMP_FILES

SamuelLarkin commented on lib/portage_utils.py:
```diff
@@ -18,27 +18,17 @@
 else:
-   from builtins import open as builtin_open
+   import builtins
+from subprocess import Popen, PIPE
```
These imports are duplicates of line 25.

SamuelLarkin committed:
Removed duplicate imports.

SamuelLarkin commented on tests/portage_utils.py/open_newlines_encodings.py:
```diff
@@ -0,0 +1,58 @@
+                )
+    with open(outfile, "wb") as f:
+        f.write(b"CR\rCRLF\r\nLF\n")
+    time.sleep(0.1)
```
A delay of 0.1 wasn't sufficient for me on GPSC.

joanise replied:
Just tested, and for me neither. 0.2 seems to be enough. Shall we just bump it?

joanise replied:
Meh, with `./run-all-tests.sh -j 4`, I needed 0.3 to pass. It's getting annoyingly long, just for one slow filesystem...

SamuelLarkin commented on lib/portage_utils.py:
```diff
@@ -184,36 +228,50 @@ def open(filename, mode='r', quiet=True, encoding="utf8"):
 
    if filename == "-":
       if mode in ('r', 'rt'):
+         theFile = builtins.open(sys.stdin.fileno(), mode, encoding=encoding, newline=newline)
```
Are we reopening the stream to control its encoding & newline?  If so, shouldn't we consider `reconfigure()`?
```
help(sys.stdin.reconfigure)
Help on built-in function reconfigure:

reconfigure(*, encoding=None, errors=None, newline=None, line_buffering=None, write_through=None) method of _io.TextIOWrapper instance
    Reconfigure the text stream with new parameters.

    This also does an implicit stream flush.
```

joanise replied:
`reconfigure` was added to Python 3.7. It would be a nicer choice, but since we don't want to lock the version of python, we have to use this older idiom supported by more versions. `io.TextIOWrapper` also does not work.

joanise committed:
make open_newlines_encodings.py more robust on GPSC

 - put the output files on /tmp, so it's a local filesystem, and thus faster
 - sleep 0.2s instead of 0.1, because otherwise the test suite sometimes fails on GPSC.

joanise committed:
Document why portage_utils.open reopens stdin/stdout
